Temporarily switch Via to be a Debian Slim-based container

robertknight opened the pull request:
This PR works around a limitation in Musl libc's DNS resolver, lack of support for TCP retries, by temporarily switching the base image of Via to be based on Debian Slim (which uses glibc instead of Musl) rather than Alpine Linux. This limitation caused problems when Via tried to connect to certain Sharepoint-hosted domains, and after some experimentation/exploration, I concluded that changing the base image in the short term is probably going to be the least unpleasant/risky solution (aside from deciding not to fix the problem at all).

See https://github.com/hypothesis/product-backlog/issues/1409#issuecomment-1338447951 for a detailed explanation of the issue and possible solutions.

Musl is due to get support for TCP retries in the next release (v1.2.4?), and will hopefully appear in Alpine 3.18. So we should be able to switch Via back to Alpine at that point, for consistency with our other apps. The Alpine-based image is also smaller. The compressed Debian-based one is ~107MB whereas the compressed Alpine-based image is 58MB.

The first commit updates the Docker file. The second adds a `run-docker` command to assist with testing locally.

Fixes https://github.com/hypothesis/product-backlog/issues/1409.

**Testing:**

1. Run h, viahtml and the client locally
2. Check out this branch, and create a local commit that patches `conf/nginx/includes/resolver.conf` to change the DNS resolver from AWS's resolver to 1.1.1.1 or 8.8.8.8
3. Run `make docker`
4. Run `make run-docker`
5. Visit http://localhost:9083/https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf

robertknight committed:
Temporarily switch Via Docker image to a Debian Slim base

Work around Musl libc's lack of support for TCP retry in DNS lookups by
temporarily switching the Via image to be based on Debian Slim. TCP
retries are expected to be added in the next Musl libc release, so at
that point it should be possible to switch back.

The compressed Alpine-based image was 58MB in size. The compressed
Debian Slim-based image is currently 107MB in size.

This resolves https://github.com/hypothesis/product-backlog/issues/1409.

robertknight committed:
Add `run-docker` command to Makefile

This is the minimal configuration needed to run a Via container locally
for testing. The h, client and viahtml services are also needed.
Checkmate is not required it seems.

In addition to these changes it is also necessary to change nginx's
resolver in conf/nginx/includes/resolver.conf to something other than
AWS's resolver.

seanh commented on Dockerfile:
```diff
@@ -1,24 +1,29 @@
-RUN apk add --no-cache nginx gettext supervisor
+RUN apt-get update && apt-get install --yes nginx nginx-extras gettext-base \
+  && pip install --no-cache-dir supervisor \
+  && apt-get clean
```
Could `supervisor` be added to the `apt-get install` command, instead of having a separate `pip` install command for it? That seems more equivalent to what used to happen with Alpine (using `apk` to install `supervisor`). `pip` may get a more recent version than `apt` does but Supervisor isn't updated often.

This was already wrong in the original Alpine-based `Dockerfile` anyway: we should be creating a virtualenv and installing a pinned version of Supervisor from a requirements file like we do with the rest of our Python dependencies. But that's a separate issue.

seanh reviewed (approved):
:+1:

This looks fine to me. I think honestly if I ever get round to doing a review of our Docker containers (and their build times) we may want to switch them all over to Debian or Ubuntu anyway. Those are the most mainstream Linux distros and what the team is familiar with and uses locally. Some good advice on Docker base images for Python apps here: https://pythonspeed.com/articles/base-image-python-docker-images/

Apparently the compressed size of [python:3.8.15-slim-bullseye](https://hub.docker.com/layers/library/python/3.8.15-slim-bullseye/images/sha256-3778adf3c19e35a6cbb7e7ef85c8715e4062f0292e7e8f876a84cc90657d2126) is only 45MB. The compressed the size is presumably what matters for uploading to and downloading from Docker Hub. But perhaps the uncompressed size affects the time taken to build the compressed image in the first place and/or the time taken to start a container?

robertknight committed:
Use supervisord as packaged by Debian

This pins the version of supervisor that we use to whatever comes with
the Debian release we are using, which is how we had things set up when
using Alpine.

robertknight commented:
> Apparently the compressed size of [python:3.8.15-slim-bullseye](https://hub.docker.com/layers/library/python/3.8.15-slim-bullseye/images/sha256-3778adf3c19e35a6cbb7e7ef85c8715e4062f0292e7e8f876a84cc90657d2126) is only 45MB. The compressed the size is presumably what matters for uploading to and downloading from Docker Hub. But perhaps the uncompressed size affects the time taken to build the compressed image in the first place and/or the time taken to start a container?

The compressed size affects upload/downloads from Docker Hub. The uncompressed size will affect anything that involves operating on the bytes of the image. How much of a practical difference it will make in our PR builds and deployment pipeline I'm not sure. There are other factors too such as apk vs apt.

Alpine has served us well for 7+ years, so I'm in no rush to change it everywhere. From what I gather in my reading, what has changed in that time is the availability of slim Ubuntu/Debian images that are more competitive with Alpine size-wise.
